File: kotlin/src/main/kotlin/y2023/Day1.kt
package y2023

import Solution
import util.NumberUtil

class Day1 : Solution {

    override fun solve(input: String): String {
        val lines = input.lines()

        fun intMerge(vals: List<Int>) = vals.first() * 10 + vals.last()

        val p1 = lines.map { line ->
            line.filter { it.isDigit() }.map { it.digitToInt() }
        }.sumOf(::intMerge)


        val p2 = lines.map { line ->
            NumberUtil.DigitMap.entries
                .fold(line) {l, e -> l.replace(e.key, e.key + e.value + e.key)}
                .filter { it.isDigit() }
                .map { it.digitToInt() }
        }.sumOf(::intMerge)

        return "Part1: $p1\nPart2: $p2"
    }
}

File: scala/src/main/scala/y2023/Day1.scala
package io.vltmn.aoc
package y2023

import util.NumberUtils.DigitMap

class Day1 extends Solution {
  def sumLine(v: Seq[Seq[Int]]): Int = v.map(l => l.head * 10 + l.last).sum

  override def solve(input: String): String = {
    val l = input.linesIterator.toSeq

    val lInts = l.map(_.toCharArray.filter(_.isDigit)
      .map(_.toInt - 48).toSeq
    )
    val p1 = sumLine(lInts)

    val lletters = l.map(line => DigitMap.toSeq
        .foldLeft(line)((acc, curr) => acc.replaceAll(curr._1, curr._1 + curr._2 + curr._1))
      .filter(_.isDigit).map(_.toInt - 48))

    val p2 = sumLine(lletters)


    s"Part1: $p1\nPart2: $p2"
  }
}

